crates/ra_hir/src/code_model.rs

```diff
@@ -1019,3 +1023,30 @@ impl Container {
         }
     }
 }
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum AssocItem {
+    Function(Function),
+    Const(Const),
+    TypeAlias(TypeAlias),
+}
+
+impl From<TraitItem> for AssocItem {
+    fn from(t: TraitItem) -> Self {
+        match t {
+            TraitItem::Function(f) => AssocItem::Function(f),
+            TraitItem::Const(c) => AssocItem::Const(c),
+            TraitItem::TypeAlias(t) => AssocItem::TypeAlias(t),
+        }
+    }
+}
+
+impl From<crate::ImplItem> for AssocItem {
+    fn from(i: crate::ImplItem) -> Self {
+        match i {
+            crate::ImplItem::Method(f) => AssocItem::Function(f),
+            crate::ImplItem::Const(c) => AssocItem::Const(c),
+            crate::ImplItem::TypeAlias(t) => AssocItem::TypeAlias(t),
+        }
+    }
+}
```

Comment:
I wonder if there's a better encoding here than tree identical enums... Perhaps we only need `AssocItem`, and can kill trait and impl items?

Reply:
Maybe... with `TraitItem` and `ImplItem`, we could additionally ensure that they are actually in a trait resp. impl. I don't know if that's worth it though... And we could also make them simple newtypes over `AssocItem`... :thinking:

Reply:
I'd rather just use `AssocItem` everywhere.